r/aws_vpc_endpoint_subnet_association: Fix issue with concurrent calls to 'ModifyVpcEndpoint'

ewbankkit opened the pull request:
Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/3382.
Requires both a lock to prevent multiple concurrent calls to `ModifyVpcEndpoint` and a delay between the calls to ensure success.

Acceptance tests:
```
make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsVpcEndpointSubnetAssociation_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAwsVpcEndpointSubnetAssociation_basic -timeout 120m
=== RUN   TestAccAwsVpcEndpointSubnetAssociation_basic
--- PASS: TestAccAwsVpcEndpointSubnetAssociation_basic (113.23s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	113.246s
```
```
make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsVpcEndpointSubnetAssociation_multiple'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAwsVpcEndpointSubnetAssociation_multiple -timeout 120m
=== RUN   TestAccAwsVpcEndpointSubnetAssociation_multiple
--- PASS: TestAccAwsVpcEndpointSubnetAssociation_multiple (250.18s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	250.191s
```

bflad commented on aws/resource_aws_vpc_endpoint_subnet_association.go:
```diff
@@ -46,10 +48,25 @@ func resourceAwsVpcEndpointSubnetAssociationCreate(d *schema.ResourceData, meta
+	// See https://github.com/terraform-providers/terraform-provider-aws/issues/3382.
+	// Prevent concurrent subnet association requests and delay between requests.
+	mk := "vpc_endpoint_subnet_association_" + endpointId
+	awsMutexKV.Lock(mk)
```
I believe we prefer to implement solutions to this problem with [`resource.Retry()`](https://godoc.org/github.com/hashicorp/terraform/helper/resource#Retry) instead of a mutex for a few reasons:
* Handles cases where the concurrent modification might be happening outside Terraform
* Better handles EC2 eventual consistency as we do not want to guess about the potential delay (it might be shorter than a minute for the API call to work)
* Slightly simpler to understand logic (you just simply declare what error(s) you are retrying on with a timeout)

What do you think?

ewbankkit replied:
I'll have to refresh my memory of what the underlying EC2 API's behavior is in this case.
I _think_ the behavior was that the `ModifyVpcEndpoint` API call just hangs waiting for a response in these situations.

bflad replied:
Okay, it'd definitely be good to check the behavior on that from the debug logs. If that is the case, where I'm guessing the SDK is automatically retrying when it probably shouldn't, we should probably fix that in by adding a handler in `config.go`. 😄

ewbankkit replied:
When I manually perform the test steps in the AWS console I notice that the VPC Endpoint goes into a _pending_ state while the subnets are being associated. The code doesn't take that into consideration. I'm going to add a `WaitForState` call and see if that improves things.

ewbankkit commented:
Adding a call to `WaitForState` did not resolve the concurrent calls issue.
However, when I look in CloudTrail at the `ModifyVpcEndpoint` calls that keep retrying I see single event like

```javascript
{
    "eventVersion": "1.05",
    "userIdentity": {
        "type": "IAMUser",
        "principalId": "XXXXXXXXXXXXXXXXXXXXX",
        "arn": "arn:aws:iam::000000000000:user/kit",
        "accountId": "000000000000",
        "accessKeyId": "XXXXXXXXXXXXXXXXXXXXX",
        "userName": "kit"
    },
    "eventTime": "2018-07-13T18:32:34Z",
    "eventSource": "ec2.amazonaws.com",
    "eventName": "ModifyVpcEndpoint",
    "awsRegion": "us-east-2",
    "sourceIPAddress": "96.255.71.4",
    "userAgent": "aws-sdk-go/1.14.24 (go1.9.2; linux; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.11.7",
    "errorCode": "Server.Unavailable",
    "requestParameters": {
        "ModifyVpcEndpointRequest": {
            "AddSubnetId": {
                "tag": 1,
                "content": "subnet-3cf40270"
            },
            "VpcEndpointId": "vpce-04a7f5432dbef0961"
        }
    },
    "responseElements": null,
    "requestID": "e967ccf1-763c-41f7-8fc2-f5a203faa2a4",
    "eventID": "1b215c5f-9d67-4170-8b4e-ffad1c0503b7",
    "eventType": "AwsApiCall",
    "recipientAccountId": "000000000000"
}
```

followed by a sequence of events like

```javascript
{
    "eventVersion": "1.05",
    "userIdentity": {
        "type": "IAMUser",
        "principalId": "XXXXXXXXXXXXXXXXXXXXX",
        "arn": "arn:aws:iam::000000000000:user/kit",
        "accountId": "000000000000",
        "accessKeyId": "XXXXXXXXXXXXXXXXXXXXX",
        "userName": "kit"
    },
    "eventTime": "2018-07-13T18:32:35Z",
    "eventSource": "ec2.amazonaws.com",
    "eventName": "ModifyVpcEndpoint",
    "awsRegion": "us-east-2",
    "sourceIPAddress": "96.255.71.4",
    "userAgent": "aws-sdk-go/1.14.24 (go1.9.2; linux; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.11.7",
    "errorCode": "Server.InternalError",
    "errorMessage": "An internal error has occurred",
    "requestParameters": {
        "ModifyVpcEndpointRequest": {
            "AddSubnetId": {
                "tag": 1,
                "content": "subnet-3cf40270"
            },
            "VpcEndpointId": "vpce-04a7f5432dbef0961"
        }
    },
    "responseElements": null,
    "requestID": "8bebfc56-2b7c-40d1-a234-1624533b62a2",
    "eventID": "a2764df4-e1ff-479f-9615-64478665c0e8",
    "eventType": "AwsApiCall",
    "recipientAccountId": "000000000000"
}
```

I'll investigate the handler suggestion.

ewbankkit commented:
The problem with adding something like

```go
	client.ec2conn.Handlers.Retry.PushBack(func(r *request.Request) {
		if r.Operation.Name == "ModifyVpcEndpoint" {
			// HTTP 503 Server.Unavailable
			// HTTP 500 Server.InternalError
			if isAWSErr(r.Error, "Unavailable", "") || isAWSErr(r.Error, "InternalError", "") {
				r.Retryable = aws.Bool(false)
			}
		}
	})
```

to `config.go` is that even when the `Server.Unavailable` or `Server.InternalError` is propagated back up to the resource's code, phantom ENIs have been created in the VPC and some of them are [cannot be deleted](https://forums.aws.amazon.com/thread.jspa?threadID=276646).

ewbankkit commented:
Rebased to remove conflict, made acceptance tests region-agnostic, added configurable timeouts to `aws_vpc_endpoint` and `aws_vpc_endpoint_subnet_association` and ensured that the VPC endpoint is back in _available_ state after modifying associated subnets.

Acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSVpcEndpoint_ -timeout 120m
=== RUN   TestAccAWSVpcEndpoint_importBasic
--- PASS: TestAccAWSVpcEndpoint_importBasic (29.30s)
=== RUN   TestAccAWSVpcEndpoint_gatewayBasic
--- PASS: TestAccAWSVpcEndpoint_gatewayBasic (27.80s)
=== RUN   TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
--- PASS: TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy (49.50s)
=== RUN   TestAccAWSVpcEndpoint_interfaceBasic
--- PASS: TestAccAWSVpcEndpoint_interfaceBasic (33.21s)
=== RUN   TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
--- PASS: TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup (206.40s)
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSService
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSService (273.67s)
=== RUN   TestAccAWSVpcEndpoint_removed
--- PASS: TestAccAWSVpcEndpoint_removed (26.15s)
PASS
```

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcEndpointSubnetAssociation_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSVpcEndpointSubnetAssociation_ -timeout 120m
=== RUN   TestAccAWSVpcEndpointSubnetAssociation_basic
--- PASS: TestAccAWSVpcEndpointSubnetAssociation_basic (211.91s)
=== RUN   TestAccAWSVpcEndpointSubnetAssociation_multiple
--- PASS: TestAccAWSVpcEndpointSubnetAssociation_multiple (515.38s)
PASS
```

ewbankkit commented:
If we don't want to merge this PR (waiting for Amazon to fix upstream), the non-concurrency related modifications are probably worth splitting into their own PR.

bflad commented:
Hey @ewbankkit I think we would like to get this PR in with all of its goodies, but I'm confused if its okay as-is given your last comment. 😄  Thanks for all your investigative work here.

ewbankkit commented:
@bflad Yes, this PR is OK as-is.

bflad commented:
This has been released in [version 1.28.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1280-july-18-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
